File: source/phpfiles/utils/weha/categorizer/TypoCorrection.php
<?php
require_once('AbstractEditAction.php');

class TypoCorrection extends AbstractEditAction {

	public function __construct($b = null) {
		$this->be = array();
		$this->be[] = $b;
	}
	
	public function isAction($edit) {
		
		if (!($edit instanceof Replacement))
		{
			return false;
		}
		else
		{
			$replEdit = $edit;
			$insContent = $replEdit->getInsertedContent();
			$delContent = $replEdit->getDeletedContent();
			
			if (count($insContent) * 10 < count($delContent) ||
				count($delContent) * 10 < count($insContent) )
				return false;
			
			$insString = "";
			$delString = "";
			
			for ($i = 0; $i < count($insContent); $i++)
				$insString .= $insContent[$i]->displayString;
			
			for ($i = 0; $i < count($delContent); $i++)
				$delString .= $delContent[$i]->displayString;
			
			if (levenshtein($insString, $delString) < 
				max((strlen($insString) + strlen($delString)) * 0.2, 3))
				return true;
		}
		
		return false;
	}

	public function classify(&$editList) {
		$ret = array();
		$newEditList = array();
		
		for ($i = 0; $i < count($editList); $i++)
		{
			$b = $editList[$i];
			if ($this->isAction($b))
				$ret[] = new TypoCorrection($b);
			else {
				$newEditList[] = $b;
			}
		}
		
		$editList = $newEditList;
		
		return $ret;
	}

}

File: weha/source/weha/wikianalysis/categorizer/TypoCorrection.java
package mo.umac.wikianalysis.categorizer;

import java.util.ArrayList;
import java.util.List;

import mo.umac.wikianalysis.diff.token.BasicEdit;
import mo.umac.wikianalysis.diff.token.Replacement;
import mo.umac.wikianalysis.lexer.WikiToken;
import mo.umac.wikianalysis.util.LevenshteinDistance;

public class TypoCorrection extends AbstractEditAction {

	{
		weight = 0.5;
	}
	
	public TypoCorrection() {
		this.be = null;
	}
	
	public TypoCorrection(BasicEdit b) {
		this.be = new BasicEdit[1];
		this.be[0] = b;
	}
	
	public boolean isAction(BasicEdit edit) {
		
		if (!(edit instanceof Replacement))
		{
			return false;
		}
		else
		{
			Replacement replEdit = (Replacement) edit;
			WikiToken[] insContent = replEdit.getInsertedContent();
			WikiToken[] delContent = replEdit.getDeletedContent();
			
			if (insContent.length > 100 ||
				delContent.length > 100 ||
				insContent.length * 10 < delContent.length ||
				delContent.length * 10 < insContent.length)
				return false;
			
			StringBuffer insString = new StringBuffer();
			StringBuffer delString = new StringBuffer();
			
			for (int i = 0; i < insContent.length; i++)
				insString.append(insContent[i].displayString);
			
			for (int i = 0; i < delContent.length; i++)
				delString.append(delContent[i].displayString);
			
			if (LevenshteinDistance.compute(insString.toString(), delString.toString()) < 
					Math.max((insString.length() + delString.length()) * 0.2, 3))
				return true;
		}
		
		return false;
	}

	@Override
	public List<AbstractEditAction> classify(List<BasicEdit> editList) {
		ArrayList<AbstractEditAction> ret = new ArrayList<AbstractEditAction>();
		ArrayList<BasicEdit> newEditList = new ArrayList<BasicEdit>();
		
		for (int i = 0; i < editList.size(); i++)
		{
			BasicEdit b = editList.get(i);
			if (this.isAction(b))
				ret.add(new TypoCorrection(b));
			else {
				newEditList.add(b);
			}
		}
		
		editList.clear();
		editList.addAll(newEditList);
		
		return ret;
	}

	@Override
	public int lengthCount() {
		return 1;
	}

}
